components/Blueprints/Validator/BlueprintValidator.php

```diff
@@ -0,0 +1,573 @@
+                    $errors[] = new ValidationError('/wordpressVersion', 'invalid-wp-version-constraint', 'Invalid WordPress version constraint.', [ 'errors' => $wpConstraintErrors ]);
+                }
+            }
+        }
+
+        // If override constraint were set (not via this class), we would validate here.
+    }
+
+    /**
+     * Summarize the blueprint.json content.
+     */
+    public static function summarizeBlueprintArray(array $blueprint): array {
+        $summ = [
+            'version' => $blueprint['version'] ?? null,
+            'constantsCount' => isset($blueprint['constants']) && is_array($blueprint['constants']) ? count($blueprint['constants']) : 0,
+            'siteOptionsCount' => isset($blueprint['siteOptions']) && is_array($blueprint['siteOptions']) ? count($blueprint['siteOptions']) : 0,
+            'muPluginsCount' => isset($blueprint['muPlugins']) && is_array($blueprint['muPlugins']) ? count($blueprint['muPlugins']) : 0,
+            'themes' => [],
+            'activeTheme' => $blueprint['activeTheme'] ?? null,
+            'plugins' => [],
+            'mediaCount' => isset($blueprint['media']) && is_array($blueprint['media']) ? count($blueprint['media']) : 0,
+            'siteLanguage' => $blueprint['siteLanguage'] ?? null,
+            'rolesCount' => isset($blueprint['roles']) && is_array($blueprint['roles']) ? count($blueprint['roles']) : 0,
+            'usersCount' => isset($blueprint['users']) && is_array($blueprint['users']) ? count($blueprint['users']) : 0,
+            'postTypesCount' => isset($blueprint['postTypes']) && is_array($blueprint['postTypes']) ? count($blueprint['postTypes']) : 0,
+            'contentCount' => isset($blueprint['content']) && is_array($blueprint['content']) ? count($blueprint['content']) : 0,
+        ];
+
+        if (!empty($blueprint['themes']) && is_array($blueprint['themes'])) {
+            foreach ($blueprint['themes'] as $themeRef) {
+                if (is_string($themeRef)) {
+                    $summ['themes'][] = $themeRef;
+                } elseif (is_array($themeRef) && isset($themeRef['source'])) {
+                    $summ['themes'][] = $themeRef['source'];
+                }
+            }
+        }
+        if (!empty($blueprint['plugins']) && is_array($blueprint['plugins'])) {
+            foreach ($blueprint['plugins'] as $pluginRef) {
+                if (is_string($pluginRef)) {
+                    $summ['plugins'][] = $pluginRef;
+                } elseif (is_array($pluginRef) && isset($pluginRef['source'])) {
+                    $summ['plugins'][] = $pluginRef['source'];
+                }
+            }
+        }
+
+        return $summ;
+    }
+
+    private function validateWpContent(Filesystem $fs): ?ValidationError {
+        if (!$fs->exists('/wp-content')) {
+            return null;
+        }
+
+        $errors = [];
+        $rootPointer = '/bundle/wp-content';
+
+        // Top-level allowed directories within wp-content
```

Comment:
I wonder if there's a reason to allow arbitrary directories 🤔 As in, themes, langs etc. would still have to be in specific paths, but would it be useful to allow, say, `my-custom-dir` here? I don't know, it's an open question – I wonder if there are sites or plugins out there that rely on those. If yes, we can adjust the spec.

Also – it's not blocking here. If anyone reports a compelling use-case later on, we can easily relax this constraint.